File: problem235.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
"""

1. find path from root to given node for p and q
2. now compare the paths from root to givven node.
3. return the last comparable node in the path. This is the lowest common ancestor
"""
class Solution:
    def lowestCommonAncestor(self, r: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        root=r
        while(root):
            if p.val>root.val and q.val>root.val:
                root = root.right
            elif p.val<root.val and q.val<root.val:
                root=root.left
            else:
                return root
            
        return root

File: problem235.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */

public class Solution {
    public TreeNode LowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        //trace path of finding both the nodes and compare the 
        //paths to find the lowest common ancestor
        List<TreeNode> path1=new List<TreeNode>(), path2=new List<TreeNode>();
        Dfs(root, p, path1);
        Dfs(root, q, path2);
        TreeNode prev=null;
        for(int i=0; i<path1.Count; i++)
        {
            if(i>=path2.Count || path1[i].val!=path2[i].val)
                return prev;
            prev=path1[i];
        }
        return prev;
    }
    
    void Dfs(TreeNode node, TreeNode x, List<TreeNode> path)
    {
        if(node==null)
            return;
        
        path.Add(node);
        
        if(node.val == x.val)
            return;
        
        if(node.val > x.val)
            Dfs(node.left, x, path);
        else
            Dfs(node.right, x, path);
        
        return;
    }
}

//time: O(log N)
//space: O(log N)